sstables/compaction_manager.cc

```diff
@@ -596,39 +596,60 @@ future<> compaction_manager::rewrite_sstables(column_family* cf, sstables::compa
     task->compacting_cf = cf;
     task->type = options.type();
     _tasks.push_back(task);
-    _stats.pending_tasks++;
 
-    task->compaction_done = repeat([this, task, options, get_func = std::move(get_func)] () mutable {
+    struct work {
+        size_t total_pending;
+        std::vector<sstables::shared_sstable> sstables;
+        work(std::vector<sstables::shared_sstable> sstables)
+            : total_pending(sstables.size())
+            , sstables(std::move(sstables))
+        {}
+    };
+
+    auto wrk = std::make_unique<work>(get_func(*cf));
+    auto w = wrk.get();
+    _stats.pending_tasks += w->total_pending;
+
+    task->compaction_done = do_for_each(w->sstables.begin(), w->sstables.end(),
+            [this, task, options, w] (sstables::shared_sstable sst) mutable {
+
         // FIXME: lock cf here
         if (!can_proceed(task)) {
-            _stats.pending_tasks--;
```

Comment:
I think it would be better to decrement `_stats.pending_tasks` here as before, instead of at the end. It is easier to keep track of all the increments/decrements.

Reply:
why do you think this is unnecessary ? If we are pushing 100 SSTables to be rewritten, there should be 100 pending tasks.

I can change it for the sake of "was like this before", but I don't think this is correct.

Reply:
I'm trying to understand what is the reason we cannot use `_stats.pending_tasks` directly to account the number of pending tasks. Why we need an additional variable for the same purpose?

Reply:
That works well if all tasks are eventually processed. The compaction manager can be stopped in which case tasks are dropped in the middle. If we rely on just `_stats.pending_tasks` we will never be able to subtract those

Reply:
In that case can I propose to replace `work` with a container and just pop sstables from it as we process them? Then it is much more clear what is happening. In the `finally()` block we can just do `_pending_task -= pending_sstables.size()`.

Reply:
I just also realized that since we use `do_for_each()`, we will have an effective concurrency of one, just before. This is even better as we don't create reactor tasks for N-1 sstables.

Reply:
I am okay with that, except I would not be able to use `do_for_each`, since popping from live containers that are iterated usually confuses `begin()` and `end()`. Maybe repeat will work, although I don't quite like the nested `repeat`

This really should've been a thread.

Reply:
nah, `do_until` works well

Reply:
`do_until([&sstables] { return sstables.empty(); }, [&sstables] {
    sst = sstables.front();
    sstables.pop_front();
    //...
});
```

Reply:
`back()` and `pop_back()` work just as well and are friendly to `std::vector`.

Reply:
yeah, I was just rusty on remembering all primitives.
I coded that already